memory/jemalloc_nodump_allocator.cc

```diff
@@ -18,250 +17,64 @@
-    // Instantiate tcache.
-    tcache_index = new unsigned(0);
-    size_t tcache_index_size = sizeof(unsigned);
-    int ret =
-        mallctl("tcache.create", tcache_index, &tcache_index_size, nullptr, 0);
-    if (ret != 0) {
-      // No good way to expose the error. Silently disable tcache.
-      delete tcache_index;
-      return MALLOCX_TCACHE_NONE;
-    }
-    tcache_.Reset(static_cast<void*>(tcache_index));
-  }
-  return MALLOCX_TCACHE(*tcache_index);
-}
-void* JemallocNodumpAllocator::Alloc(extent_hooks_t* extent, void* new_addr,
-                                     size_t size, size_t alignment, bool* zero,
-                                     bool* commit, unsigned arena_ind) {
-  extent_alloc_t* original_alloc =
-      original_alloc_.load(std::memory_order_relaxed);
-  assert(original_alloc != nullptr);
-  void* result = original_alloc(extent, new_addr, size, alignment, zero, commit,
-                                arena_ind);
-  if (result != nullptr) {
-    int ret = madvise(result, size, MADV_DONTDUMP);
-    if (ret != 0) {
-      fprintf(
-          stderr,
-          "JemallocNodumpAllocator failed to set MADV_DONTDUMP, error code: %d",
-          ret);
-      assert(false);
-    }
-  }
-  return result;
-}
-
-Status JemallocNodumpAllocator::DestroyArena(unsigned arena_index) {
-  assert(arena_index != 0);
-  std::string key = "arena." + std::to_string(arena_index) + ".destroy";
-  int ret = mallctl(key.c_str(), nullptr, 0, nullptr, 0);
-  if (ret != 0) {
-    return Status::Incomplete("Failed to destroy jemalloc arena, error code: " +
-                              std::to_string(ret));
-  }
-  return Status::OK();
-}
-
-void JemallocNodumpAllocator::DestroyThreadSpecificCache(void* ptr) {
-  assert(ptr != nullptr);
-  unsigned* tcache_index = static_cast<unsigned*>(ptr);
-  size_t tcache_index_size = sizeof(unsigned);
-  int ret __attribute__((__unused__)) =
-      mallctl("tcache.destroy", nullptr, 0, tcache_index, tcache_index_size);
-  // Silently ignore error.
-  assert(ret == 0);
-  delete tcache_index;
+Status NewJemallocNodumpAllocator(
+    JemallocAllocatorOptions& options,
+    std::shared_ptr<MemoryAllocator>* memory_allocator) {
+  return NewJemallocNodumpAllocator(options, 1 /* num_arenas */,
+                                    memory_allocator);
 }
 
-#endif  // ROCKSDB_JEMALLOC_NODUMP_ALLOCATOR
-
 Status NewJemallocNodumpAllocator(
-    JemallocAllocatorOptions& options,
+    JemallocAllocatorOptions& options, size_t num_arenas,
     std::shared_ptr<MemoryAllocator>* memory_allocator) {
   if (memory_allocator == nullptr) {
     return Status::InvalidArgument("memory_allocator must be non-null.");
   }
 #ifndef ROCKSDB_JEMALLOC
   (void)options;
+  (void)num_arenas;
   return Status::NotSupported("Not compiled with JEMALLOC");
 #else
-  std::unique_ptr<MemoryAllocator> allocator(
-      new JemallocNodumpAllocator(options));
+  std::unique_ptr<MemoryAllocator> allocator;
+  switch (num_arenas) {
```

Comment:
It's possible to avoid the switch/case with metaprogramming, but this code is relatively clear and concise.